feat(datasets) Move query splitter at at the end of query processing

fpacifici opened the pull request:
This rebuilds https://github.com/getsentry/snuba/pull/831

This PR moves the query splitting from the very beginning of the query to the very end and makes the split specifications storage specific.
In details:
- removes the split_query decorator 
- Allows the storage to specify the algorithms the storage supports to perform splitting (split option 1 as discussed). This allows us to configure splitting depending on the structure of the table and on its primary key.
- Makes splitting formally allowed only as optimizations. See docstring in the storage class
- Executes the splitting process as part of the QueryExecutionStrategy
- Remove the dependencies the splitting logic had on query extensions and make it depend on the query object.

fpacifici committed:
Move split into the Strategy abstraction

fpacifici committed:
Move splitters to storage

fpacifici committed:
Fix some imports

fpacifici committed:
Fix tests

fpacifici committed:
Make the splitter not depend on extensions

fpacifici committed:
Move file back

fpacifici committed:
Fix tests

fpacifici committed:
Add more tests

fpacifici committed:
Add some comments

fpacifici committed:
Nits

fpacifici commented on snuba/datasets/dataset.py:
```diff
@@ -153,12 +138,6 @@ def process_condition(self, condition) -> Tuple[str, str, Any]:
         """
         return condition
 
-    def get_split_query_spec(self) -> Union[None, ColumnSplitSpec]:
```
This is another step in removing stuff from the dataset that should not be there.

fpacifici commented on snuba/datasets/groups.py:
```diff
@@ -44,6 +45,9 @@ def get_table_writer(self) -> Optional[TableWriter]:
     def get_query_processors(self) -> Sequence[QueryProcessor]:
         return [SimpleJoinOptimizer(), PrewhereProcessor()]
 
+    def get_query_splitters(self) -> Sequence[StorageQuerySplitStrategy]:
+        return []
+
```
Splitting a join query is a more complex issue. We cannot simply take the splitters from the storages we are joining and apply all of them. In most cases the query would not even work.
I think the decision belongs to the JoinQueryPlanBuilder (that does not yet exists). It knows the entities we are joining, it knows which storages have been selected, it knows which type of joins we are applying, so it can decide if any split strategy can be applied.
Hopefully we will be able to make this process generic and not custom per dataset.

fpacifici commented on snuba/datasets/storages/events.py:
```diff
@@ -317,4 +322,14 @@ def get_promoted_tags() -> Mapping[str, Sequence[str]]:
+    query_splitters=[
+        ColumnSplitQueryStrategy(
+            ColumnSplitSpec(
+                id_column="event_id",
+                project_column="project_id",
+                timestamp_column="timestamp",
```
This is how the storage decides what can be split depending on the primary key.

fpacifici commented on snuba/web/query.py:
```diff
@@ -57,7 +56,6 @@ def parse_and_run_query(
     return result
 
 
-@split_query
```
I felt surprisingly well when I removed this  :D

fpacifici commented on snuba/web/split.py:
```diff
@@ -1,10 +1,16 @@
 import copy
```
There is no reason for this file to be in /web. I left it there only to make this PR easier to review since github shows the change with respect to the previous split logic instead of showing a new file. (I tried to move the file first but somehow after a few commits, github lost track the file was moved and just showed a brand new file).
Also I kept the refactoring of this class to the very minimum needed to move the splitting logic to make the PR easier to review. I think more refactoring is worth doing.

Moving the file to where it should be will be trivial

tkaemming replied:
Let's keep that as a separate change, I don't want to derail another otherwise good change with what does and doesn't belong in web again. 😞

fpacifici replied:
Unfortunately, if I move the file here (I tried), github will just show you a brand new file with no connection to the previous one (still don't get why, it should understand I did a git mv).
That would make it really hard to spot what actually changed in the split code which is already complex.

fpacifici replied:
https://github.com/getsentry/snuba/pull/883

tkaemming commented on snuba/web/split.py:
```diff
@@ -128,9 +173,51 @@ def time_split(dataset, request: Request, *args, **kwargs):
-        dataset, request: Request, column_split_spec: ColumnSplitSpec, *args, **kwargs
-    ):
+
+class ColumnSplitSpec(NamedTuple):
+    """
+    Provides the column names needed to perform column splitting.
+    id_column represent the identity of the row.
+
+    """
+
+    id_column: str
+    project_column: str
+    timestamp_column: str
+
+    def get_min_columns(self) -> Sequence[str]:
+        return [self.id_column, self.project_column, self.timestamp_column]
```
It seems like this data structure is just used as a constructor parameter to `ColumnSplitQueryStrategy` at this point. Should these just be instance attributes on that class now? Or is there other benefit to keeping this around?

fpacifici replied:
There is no benefit in keeping it around. In this case as well I kept it only not to touch how the splitter worked internally and focus on the change to the code structure to make the change in the PR more evident. 
Though I am ok removing it right away.

tkaemming commented on snuba/web/split.py:
```diff
@@ -128,9 +173,51 @@ def time_split(dataset, request: Request, *args, **kwargs):
+
+        selected_cols = request.query.get_selected_columns()
+        return (
+            selected_cols is not None
+            and len(selected_cols) > 0
+            and not request.query.get_aggregations()
```
This all seems like it should be checked before the `copy` and early return, foregoing the copy if not required.

tkaemming commented on snuba/datasets/plans/single_storage.py:
```diff
@@ -18,7 +19,16 @@
+
     def execute(self, request: Request, runner: QueryRunner) -> RawQueryResult:
+        for splitter in self.__splitters:
+            if splitter.can_execute(request):
+                return splitter.execute(request, runner)
```
Would it make sense to skip `can_execute` and directly call the splitter (to avoid things like the double request copy in `ColumnSplitQueryStrategy`), letting it return `None` it if it is unable to run the query, indicating that another strategy should be used?

fpacifici replied:
Yes it is possible, but is it better? Feels a little implicit that we call `execute` and depending on the result we figure out that nothing was executed and we have to try again with another splitter. Anyway, not a strong opinion.

tkaemming replied:
Other than potentially (and admittedly minor) efficiency improvements, the other benefit would be that it reduces the potential for this precondition to be violated: https://github.com/getsentry/snuba/pull/865#discussion_r404552098 Once that type becomes immutable (at least via the type definition — assuming it will) that's less of a concern from my perspective. I think either approach right now has it's pros and cons, I'll leave the decision up to you.

fpacifici replied:
good point, mutability can be an issue, will make the change.

tkaemming commented on snuba/web/split.py:
```diff
@@ -13,39 +19,65 @@
+            and has_from_date
+            and has_to_date
+            and orderby[:1] == [f"-{self.__timestamp_col}"]
+            and remaining_offset < 1000
```
Similar idea here, checking whether or not an integer is within range seems quite a bit more efficient than the generator construction that happens for the conditions checks.

tkaemming reviewed (changes requested):
Overall direction looks really good, there are some implementation concerns and nits inline though.

fpacifici commented on snuba/datasets/plans/split_strategy.py:
```diff
@@ -0,0 +1,21 @@
+    to select the valid algorithm.
+    """
+
+    @abstractmethod
+    def can_execute(self, request: Request) -> bool:
+        """
+        Returns True if this split algorithm can be applied to the Query provided.
+        """
+        raise NotImplementedError
```
Indeed,
- I will change the comment since I don't know how long the request will remain there
- yes it is correct, request will be replaced by the storage query.

tkaemming replied:
Cool, thanks for clarifying.

fpacifici commented on snuba/web/split.py:
```diff
@@ -13,39 +19,65 @@
-        query_limit = request.query.get_limit()
-        limit = query_limit if query_limit is not None else 0
+def _is_query_splittable(query: Query) -> bool:
+    (use_split,) = state.get_configs([("use_split", 0)])
```
I wonder whether we should get rid of the "use_split" config at all. It seems it was some sort of killswitch flag for the rollout.

tkaemming replied:
I'd check with @JTCunning but I'm fine with removing it. @lynnagara discovered a while back that it had been broken for months and nobody noticed: https://github.com/getsentry/snuba/pull/739#discussion_r374402603

fpacifici commented on snuba/web/split.py:
```diff
@@ -13,39 +19,65 @@
+        self.__timestamp_col = timestamp_col
+
+    def can_execute(self, request: Request) -> bool:
         remaining_offset = request.query.get_offset()
```
right, will rename

fpacifici commented on snuba/web/split.py:
```diff
@@ -13,39 +19,65 @@
+            _is_query_splittable(request.query)
+            and has_from_date
+            and has_to_date
+            and orderby[:1] == [f"-{self.__timestamp_col}"]
```
right, I will do a cleanup pass on the split algorithm code. 
Initially I wanted to touch the existing code as little as possible to avoid making reviewers to  cope with changes to the splitting algorithm. But by removing the dependency on the extensions it seems the algorithm was changed anyway so it may look confusing now.

fpacifici commented on snuba/web/split.py:
```diff
@@ -65,32 +97,45 @@ def time_split(dataset, request: Request, *args, **kwargs):
-            request.extensions["timeseries"]["to_date"], date_align
+        to_date_str = next(
+            condition[2]
+            for condition in request.query.get_conditions() or []
+            if _identify_condition(condition, self.__timestamp_col, "<")
         )
-        from_date = util.parse_datetime(
-            request.extensions["timeseries"]["from_date"], date_align
+        from_date_str = next(
+            condition[2]
+            for condition in request.query.get_conditions() or []
+            if _identify_condition(condition, self.__timestamp_col, ">=")
         )
```
makes sense

lynnagara commented on snuba/web/split.py:
```diff
@@ -128,9 +173,51 @@ def time_split(dataset, request: Request, *args, **kwargs):
+
+class ColumnSplitSpec(NamedTuple):
+    """
+    Provides the column names needed to perform column splitting.
```
It's also used for the time splitting not just column? (in the case of timestamp_column)

fpacifici committed:
Merge branch 'master' into feat/storage_splitter

fpacifici committed:
Nits

fpacifici committed:
Remove can_execute

fpacifici committed:
Clean up split method

fpacifici committed:
Fix comment

tkaemming commented on snuba/web/split.py:
```diff
@@ -57,40 +90,53 @@ def time_split(dataset, request: Request, *args, **kwargs):
-
-        to_date = util.parse_datetime(
-            request.extensions["timeseries"]["to_date"], date_align
+        # We already ensured the conditions below exist in the query during the
+        # can_execute call.
```
Seems like it'd be better to just find them once then, now that there is no split between `execute` and `__can_execute` from the caller's perspective.

fpacifici replied:
Fair point, I forgot about that.

tkaemming commented on snuba/web/split.py:
```diff
@@ -57,40 +90,53 @@ def time_split(dataset, request: Request, *args, **kwargs):
         )
+        from_date = util.parse_datetime(from_date_str, date_align)
+
+        limit = request.query.get_limit() or 0
```
This `or` is redundant with the check in `_is_query_splittable` as well.

fpacifici replied:
It is still necessary for mypy not to complain.

tkaemming replied:
This also applies to https://github.com/getsentry/snuba/pull/865/files#r410393313 but writing it here since this is a bit more self-contained an example. This seems more of a reflection of the structure of this method and the duplicated logic in relationship to `__can_execute` and `_is_query_splittable` than of deficiencies in the type checker.

If `execute` was structured this way, this wouldn't be an issue:

```python
def execute(…) -> Optional[RawQueryResult]:
  # establish invariants (basically inline __can_execute and _is_query_splittable)
  limit = request.query.get_limit()
  if limit is None:
    return None

  to_date = …
  if to_date is None:
    return None

  # run split
  … everything else already in here …
```

This was the point of this comment but I guess it was more implied than directly called out: https://github.com/getsentry/snuba/pull/865#discussion_r404557527

tkaemming commented on snuba/web/split.py:
```diff
@@ -128,61 +174,100 @@ def time_split(dataset, request: Request, *args, **kwargs):
+            return False
+
+        total_col_count = len(request.query.get_all_referenced_columns())
+        copied_query = copy.deepcopy(request.query)
```
This copies twice now. 😞

fpacifici replied:
The next statement uses the copy and modifies it to decide whether we can split.
I can move this into the execute method instead. But the copy is needed to decide whether the splitter can be executed.

tkaemming replied:
The end result of the copy and `set_selected_columns` is the same as line 232-233, isnt it?

The double copy issue was also mentioned here: https://github.com/getsentry/snuba/pull/865#discussion_r404557527

fpacifici replied:
Well, I think it is better to just remove the can_execute methods

tkaemming commented on snuba/web/split.py:
```diff
@@ -128,61 +174,100 @@ def time_split(dataset, request: Request, *args, **kwargs):
-            request.extensions["project"]["project"] = project_ids
 
-            timestamp_field = column_split_spec.timestamp_column
+            timestamp_field = self.__timestamp_column
```
Odd that this is assigned to a local variable but `self.__project_column` isn't above which also is used in a loop on the result payload in a nearly identical fashion.

fpacifici replied:
Will change

tkaemming commented on snuba/web/split.py:
```diff
@@ -128,61 +174,100 @@ def time_split(dataset, request: Request, *args, **kwargs):
             )
+            request.query.add_conditions([(self.__id_column, "IN", event_ids)])
             request.query.set_offset(0)
             request.query.set_limit(len(event_ids))
```
Just noticed this assumes the `event_id` is globally unique rather than unique per project, so this logic is incorrect but probably not often enough to be noticable.

fpacifici replied:
Added a TODO.

lynnagara replied:
Good point. I don't recall why that `set()` was needed.

tkaemming reviewed (approved):
This works, I still have concerns about the split implementations themselves but nothing major enough to be blocking, I guess.

fpacifici commented on snuba/web/split.py:
```diff
@@ -13,39 +19,66 @@
 
-def split_query(query_func):
-    def wrapper(dataset, request: Request, *args, **kwargs):
+def _is_query_splittable(query: Query) -> bool:
+    query_limit = query.get_limit()
+    return query_limit is not None and query_limit > 0 and not query.get_groupby()
```
Could not find an answer, I brought it from the previous implementation

fpacifici commented on snuba/web/split.py:
```diff
@@ -13,39 +19,66 @@
+            cond
+            if not _identify_condition(cond, field, operand)
+            else [field, operand, new_literal]
+            for cond in query.get_conditions() or []
```
There was a case early in the query processing that needed this. It may not be the case anymore now.

fpacifici commented on snuba/web/split.py:
```diff
@@ -13,39 +19,66 @@
+        self.__timestamp_col = timestamp_col
+
+    def __can_execute(self, request: Request) -> bool:
         (use_split,) = state.get_configs([("use_split", 0)])
```
I thought this addressed that comment, thus I am not sure I get where you suggest this should happen.

tkaemming replied:
Whether splitting is enabled _at all_ seems like it should be a question that is answered by either the plan builder or the execution strategy, not the splitter instances themselves.

fpacifici commented on snuba/web/split.py:
```diff
@@ -57,40 +90,53 @@ def time_split(dataset, request: Request, *args, **kwargs):
+        # can_execute call.
+        to_date_str = next(
+            condition[2]
+            for condition in request.query.get_conditions() or []
```
It still is necessary because mypy does not know that we verified get_conditions cannot be None.

fpacifici committed:
Merge branch 'master' into feat/storage_splitter

fpacifici committed:
Nits

tkaemming commented on snuba/web/split.py:
```diff
@@ -57,40 +59,68 @@ def time_split(dataset, request: Request, *args, **kwargs):
+        if limit is None or request.query.get_groupby():
+            return None
+
+        if request.query.get_offset() >= 1000:
+            return None
```
I wonder what this is/was for.

fpacifici replied:
could not find an answer to this.

tkaemming commented on snuba/web/split.py:
```diff
@@ -57,40 +59,68 @@ def time_split(dataset, request: Request, *args, **kwargs):
-            # evaluation, so we need to copy the body to ensure that the query
-            # has not been modified in between this call and the next loop
-            # iteration, if needed.
-            # XXX: The extra data is carried across from the initial response
-            # and never updated.
```
This is still true, nothing has changed. If we're going to remove this reminder it should go on the debt list.

This comment wasn't fixed, but marked as resolved. https://github.com/getsentry/snuba/pull/865#discussion_r404553324

fpacifici replied:
To be fair I removed the comment because I found it more confusing than helpful, (thus also marked the comment as resolved). I can bring it back if you find useful.

tkaemming replied:
I'll put it on the debt list but I don't understand that thought process to be honest.

fpacifici replied:
Will rewrite it with more details to be more explicit.

fpacifici replied:
added the comment back with more details.

tkaemming commented on snuba/web/split.py:
```diff
@@ -128,61 +158,87 @@ def time_split(dataset, request: Request, *args, **kwargs):
+        self.__project_column = project_column
+        self.__timestamp_column = timestamp_column
+
+    def __get_min_columns(self) -> Sequence[str]:
+        return [self.__id_column, self.__project_column, self.__timestamp_column]
```
This is only used once now, I'm not sure it's beneficial to make it its own method.

tkaemming commented on snuba/web/split.py:
```diff
@@ -128,61 +158,87 @@ def time_split(dataset, request: Request, *args, **kwargs):
+    ) -> Optional[RawQueryResult]:
         """
         Split query in 2 steps if a large number of columns is being selected.
             - First query only selects event_id and project_id.
```
Not new, but this comment is wrong since (it also includes the timestamp.)

fpacifici replied:
ok

tkaemming commented on snuba/web/split.py:
```diff
@@ -57,40 +59,68 @@ def time_split(dataset, request: Request, *args, **kwargs):
         while split_start < split_end and total_results < limit:
-            request.extensions["timeseries"]["from_date"] = split_start.isoformat()
-            request.extensions["timeseries"]["to_date"] = split_end.isoformat()
+            # We need to make a copy to use during the query execution because we replace
+            # the start-end conditions on the query at each iteration of this loop.
+            split_request = copy.deepcopy(request)
```
Processors may also mutate this too, right? The scope of mutation isn't just limited this function like this comment change implies.

fpacifici replied:
No, we are past the processors at this point. The only changes to the query are those done by this method, thus the comment.

tkaemming replied:
Sorry, my mistake, somehow I managed to forget that was the entire point of this change to begin with.

fpacifici committed:
Merge branch 'master' into feat/storage_splitter

fpacifici committed:
Address comments

fpacifici committed:
Replace in the prewhere clause as well

fpacifici commented:
Found an issue. The column splitter now operates after the prewhere processor. Thus with this change the second query of the column splitter would be 
```
2020-04-21 13:24:27,015 Query: SELECT event_id, timestamp, project_id, level, logger FROM sentry_local PREWHERE project_id IN (681) WHERE timestamp >= toDateTime('2020-04-21T00:0
0:00') AND timestamp < toDateTime('2020-04-21T03:50:01') AND deleted = 0 AND event_id IN ('00000000000000000000000000000000', '96009600960096009600960096009600', '180018001800180
01800180018001800', '54005400540054005400540054005400', '11400114001140011400114001140011', '12000120001200012000120001200012', '72007200720072007200720072007200', '42004200420042004200420042004200', '13800138001380013800138001380013', '12600126001260012600126001260012', '90009000900090009000900090009000', '10800108001080010800108001080010', '66006600660
066006600660066006600', '48004800480048004800480048004800', '12001200120012001200120012001200', '24002400240024002400240024002400', '84008400840084008400840084008400', '60006000600060006000600060006000', '13200132001320013200132001320013', '10200102001020010200102001020010', '60060060060060060060060060060060', '36003600360036003600360036003600', '3000300
0300030003000300030003000', '78007800780078007800780078007800') ORDER BY timestamp DESC LIMIT 0, 24      
```

while before it was like:
```
SELECT event_id, timestamp, project_id, level, logger FROM sentry_local PREWHERE event_id IN ('30003000300030003000300030003000', '6600660066006600
6600660066006600', '48004800480048004800480048004800', '12001200120012001200120012001200', '54005400540054005400540054005400', '10800108001080010800108001080010', '11400114001140
011400114001140011', '36003600360036003600360036003600', '60006000600060006000600060006000', '10200102001020010200102001020010', '13800138001380013800138001380013', '000000000000
00000000000000000000', '13200132001320013200132001320013', '72007200720072007200720072007200', '12600126001260012600126001260012', '12000120001200012000120001200012', '9600960096
0096009600960096009600', '60060060060060060060060060060060', '78007800780078007800780078007800', '90009000900090009000900090009000', '24002400240024002400240024002400', '84008400
840084008400840084008400', '42004200420042004200420042004200', '18001800180018001800180018001800') WHERE project_id IN (681) AND timestamp >= toDateTime('2020-04-21T00:00:00') AN
D timestamp < toDateTime('2020-04-21T03:50:01') AND deleted = 0 ORDER BY timestamp DESC LIMIT 0, 24
```

So need to go back to the drawing board to see how to solve this.

fpacifici committed:
Merge branch 'master' into feat/storage_splitter

fpacifici committed:
Perform query processing in split strategy

fpacifici committed:
Merge branch 'master' into feat/storage_splitter

fpacifici committed:
Adapt splitter to db query processors

fpacifici committed:
Cleanup and fix the column splitter

fpacifici commented on snuba/datasets/plans/split_strategy.py:
```diff
@@ -0,0 +1,40 @@
+
+    The main difference between this class and the QueryPlanExecutionStrategy is that
+    it relies on a smarter QueryRunner (SplitQueryRunner) than the one provided to the
+    execution strategy. The runner this class receives is supposed to take care of
+    running the DB query processors before executing the query on the database so that
```
I am still a bit nervous of having processors that should be executed right before talking to the database when a splitter is orchestrating this process.
I think, this way, by keeping the responsibility of running the processors out of the splitter:
- we unload some responsibility from the splitter class
- db query processors are entirely a matter of the same class (the execution strategy, none else knows how to run them)
- we constrain the splitter strategy not to be too "creative" 

We do not need the flexibility of letting the splitter tweak which processors are executed, and probably we should try to avoid that. In case there was a need for that, it will be a fairly small scope change.

fpacifici commented on snuba/web/split.py:
```diff
@@ -128,61 +159,89 @@ def time_split(dataset, request: Request, *args, **kwargs):
+        if not result.result["data"]:
+            return None
+
+        # Making a copy just in case runner returned None (which would drive the execution
+        # strategy to ignore the result of this splitter and try the next one).
+        query = copy.deepcopy(query)
```
This should never be needed. I added it defensively just in case we add another condition that returns None and we forget that we mutated the query.

fpacifici committed:
Merge branch 'master' into feat/storage_splitter

tkaemming commented on snuba/web/split.py:
```diff
@@ -128,61 +159,89 @@ def time_split(dataset, request: Request, *args, **kwargs):
+        result = runner(minimal_query, request_settings)
+        del minimal_query
+
+        if not result.result["data"]:
+            return None
```
I guess this is necessary because we need to actually run the query to get the column types for the full selected column set from ClickHouse?

fpacifici replied:
Yes, in theory we should still run the full query on clickhouse to honor the snuba contract. In practice I wonder if that is actually a problem for the product I think it would be worth investigating. Since we are running twice a query that returns no results.

tkaemming reviewed (approved):
This looks good.

fpacifici commented on snuba/datasets/storage.py:
```diff
@@ -71,6 +72,16 @@ def get_query_processors(self) -> Sequence[QueryProcessor]:
         """
         raise NotImplementedError
 
+    @abstractmethod
+    def get_query_splitters(self) -> Sequence[QuerySplitStrategy]:
+        """
+        If this storage supports splitting queries as optimizations, they are provided here.
+        These are optimizations, the query plan builder may decide to override the storage
+        and to skip the splitters. So correctness of the query must not depend on these
+        strategies to be applied.
+        """
+        raise NotImplementedError
```
I think it makes sense in general. Will add. Though I think it will have very little impact concerning joins because joinedStorage is supposed to die as soon as I have time to revamp this: https://github.com/getsentry/snuba/pull/830

And actually have a join query joining two storages instead of having a fake single storage that could never be plugged into an entity anyway.

fpacifici committed:
Merge branch 'master' into feat/storage_splitter

fpacifici committed:
Address last comments

fpacifici committed:
Merge branch 'feat/storage_splitter' of https://github.com/getsentry/snuba into feat/storage_splitter

fpacifici committed:
Merge branch 'master' into feat/storage_splitter
